src/NHibernate/Loader/DynamicBatchingHelper.cs

```diff
@@ -0,0 +1,58 @@
+using System;
+using NHibernate.Criterion;
+using NHibernate.Engine;
+using NHibernate.SqlCommand;
+
+namespace NHibernate.Loader
+{
+	internal class DynamicBatchingHelper
+	{
+		private static string BatchIdPlaceholder = "$$BATCH_ID_PLACEHOLDER$$";
+
+		public static SqlStringBuilder BuildBatchFetchRestrictionFragment()
+		{
+			return new SqlStringBuilder(1).Add(BatchIdPlaceholder);
+		}
+
+		public static void ExpandBatchIdPlaceholder(SqlString sqlString, QueryParameters queryParameters, string[] columns, Dialect.Dialect dialect, out SqlString result, out Parameter[] parameters)
+		{
+			var wherePart = GenerateWherePart(queryParameters, columns, dialect, out parameters);
```

Comment:
At this point you already know the number of parameters required. `parameters.Length == columns.Length * queryParameters.PositionalParameterValues.Length`. So you can pass the parameter's array without need of an out parameter.

Reply:
Then... `InExpression.GetSqlString` is always used in this combination:

```csharp
var bogusParam = Parameter.Placeholder;
var wherePart = InExpression.GetSqlString(columns, paramsCount, dialect, bogusParam);
wherePart.SubstituteBogusParameters(parameters, bogusParam);
```
This could and should be one method.

Reply:
This split is not very useful in the first place and now you are suggesting to complicate it even further by some pre-calculations based on the internal implementation details. I don't like it. Sorry I'm going to undo your last commit.

Reply:
It's 5 lines of code now - just let it be..

Reply:
I'm just re-using the pattern that's used in `Loader`. Just check how `CreateSqlCommand` works and see `GetParameterSpecifications`

Reply:
The difference between `GetParameterSpecifications` mentioned and this method that here we know the structure of the string fragment and number of parameters and can write more optimal and readable code. And in `GetParameterSpecifications` the `SqlString` could be anything. It could be a user-provided select statement or a generated string, or something else.

Reply:
You've just introduced yet another version of `CreateParameterSpecificationsAndAssignBackTrack`. I really don't see how it improves anything...